### ISSUE TITLE
c.transient_for returns nil

### ISSUE THREAD

**Issue by basaran** [awesomeWM/awesome#3390]
Hi, I'm trying get the transient_for property of a client, xprop returns this:

```
WM_TRANSIENT_FOR(WINDOW): window id # 0x180002b
```

but client.focus.transient_for returns nil. Is `c.transient_for` means something else?

**Comment by Elv13** [comment]
Hello,

Can you check all clients and see if one has a matching window id? Maybe the "other" window id in xprop isn't visible or isn't recognized as a client for some reason (like a tooltip)?

**Comment by basaran** [comment]
Hello, these are the relevant bits for cudatext, a text editor. Situation is rather odd, this is probably an inconsistency with cudatext.  Popup returns transient for `0x560002b`

```
xwininfo -tree -root

     0x56000cb "cudatext":
     0x56000c3 (has no name):
     0x560002b "cudatext": ("cudatext" "Cudatext")
        1 child:
        0x560002c (has no name): ()
     0x5600024 "cudatext": ("cudatext" "Cudatext")
        1 child:
        0x5600025 (has no name):
     0x5600003 "cudatext": ("cudatext" "Cudatext")
        1 child:
        0x5600004 (has no name):
     0x5600001 "cudatext": ("cudatext" "Cudatext")
        1 child:
        0x5600002 (has no name):
```

The popup with the transient_for attribute xprop:

```
awful.client.property.single_instance_id(UTF8_STRING) =
                window id # of group leader: 0x5600001
WM_TRANSIENT_FOR(WINDOW): window id # 0x560002b
_NET_WM_USER_TIME_WINDOW(WINDOW): window id # 0x560023c
WM_CLIENT_LEADER(WINDOW): window id # 0x5600001
_NET_WM_PID(CARDINAL) = 1483657
```

**Comment by psychon** [comment]
> The popup with the transient_for attribute xprop:

Is that everything? If so, I would guess that this is not a visible window. Or at least it is a weird window since e.g. `WM_PROTOCOLS` is pretty much always set. But that `single_instance_id` is weird... `xwininfo` could confirm (its `Map State` should be `IsViewable`).

Another guess would be a loop in the `WM_TRANSIENT_FOR` relation. Could you repeat the outputs but this time also record which window has which id (`xwininfo` can tell you that)?

Well...

or perhaps like this:
- Run `xwininfo` on the poup. Make note of the `Window id`.
- Run `xprop` on the popup and get its `WM_TRANSIENT_FOR`. I will assume it is `WM_TRANSIENT_FOR(WINDOW): window id # 0x560002b`.
- Run `xwininfo -id 0x560002b`
- Run `xprop -id 0x560002b`

**Comment by basaran** [comment]
The ids appear to be dynamic from cudatext, so they are different below than the original post.

#### xwininfo, click on popup
```
xwininfo: Window id: 0x6c0850b (has no name)
```

#### xprop, click on popup
```
WM_TRANSIENT_FOR(WINDOW): window id # 0x6c0002b
```

#### xwininfo -id 0x6c0002b (transient_for id)
```
xwininfo: Window id: 0x6c0002b "cudatext"

  Absolute upper-left X:  0
  Absolute upper-left Y:  0
  Relative upper-left X:  0
  Relative upper-left Y:  0
  Width: 1
  Height: 1
  Depth: 24
  Visual: 0x21
  Visual Class: TrueColor
  Border width: 0
  Class: InputOutput
  Colormap: 0x20 (installed)
  Bit Gravity State: NorthWestGravity
  Window Gravity State: NorthWestGravity
  Backing Store State: NotUseful
  Save Under State: no
  Map State: IsUnMapped
  Override Redirect State: no
  Corners:  +0+0  -2559+0  -2559-1439  +0-1439
  -geometry 1x1+0+0
```

#### xprop -id 0x6c0002b (transient_for id)
```

WM_STATE(WM_STATE):
                window state: Withdrawn
                icon window: 0x0
_NET_FRAME_EXTENTS(CARDINAL) = 2, 2, 10, 5
awful.client.property.single_instance_id(UTF8_STRING) =
WM_HINTS(WM_HINTS):
                Client accepts input or input focus: True
                Initial state is Normal State.
                window id # of group leader: 0x6c00001
_NET_WM_SYNC_REQUEST_COUNTER(CARDINAL) = 113246253
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_NORMAL
_NET_WM_USER_TIME(CARDINAL) = 785255115
_NET_WM_USER_TIME_WINDOW(WINDOW): window id # 0x6c0002c
WM_CLIENT_LEADER(WINDOW): window id # 0x6c00001
_NET_WM_PID(CARDINAL) = 232440
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLIENT_MACHINE(STRING) = "michael-arch"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified location: 0, 0
                program specified minimum size: 1 by 1
                window gravity: NorthWest
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, WM_TAKE_FOCUS, _NET_WM_PING, _NET_WM_SYNC_REQUEST
WM_CLASS(STRING) = "cudatext", "Cudatext"
WM_ICON_NAME(STRING) = "cudatext"
_NET_WM_ICON_NAME(UTF8_STRING) = "cudatext"
WM_NAME(STRING) = "cudatext"
_NET_WM_NAME(UTF8_STRING) = "cudatext"
```

**Comment by psychon** [comment]
Thanks!

Both of the following say that this is not a window managed by the WM. Thus, there is no `client` object for window `0x6c0002b` that could be used. I don't know why things are this way, but that's what the output says.

> Map State: IsUnMapped

> WM_STATE(WM_STATE): window state: Withdrawn

Also: This is "weird" and also does not look like a "real" window:
```
  Absolute upper-left X:  0
  Absolute upper-left Y:  0
  Relative upper-left X:  0
  Relative upper-left Y:  0
  Width: 1
  Height: 1
```

**Comment by basaran** [comment]
> This is "weird" and also does not look like a "real" window:

![image](https://user-images.githubusercontent.com/30809170/126399321-5d4f026e-35b2-4cc5-84fe-f12b60e894a2.png)

The editor is written with `lazarus` and `object pascal`, and I confirmed this transient_for problem is the same for all other popups (file save, close etc).

It appears nothing is wrong with awesomewm, and there is enough data here if anybody runs into this in the future. Closing.
